### ISSUE TITLE
Add Component.isAttached()

### ISSUE THREAD

**Issue by mvysny** [vaadin/flow#7911]
Currently in Vaadin 14.1.19 it's only possible to check whether a component is attached, by running the following method:

```kotlin
fun Component.isAttached(): Boolean {
    val ui: UI = ui.orElse(null) ?: return false
    return !ui.isClosing
}
```

It would be good to have `Component.isAttached()` directly provided by Vaadin.

**Comment by mvysny** [comment]
I wonder if a component may be perceived as attached even if the UI itself is detached from the session?

**Comment by denis-anisimov** [comment]
`Component::getElement::getNode::isAttached`.

I agree it's not obvious but I would avoid put everything into `Component`  as well: it litters API and makes it's hard to find required methods.

I would close this ticket but it's up to you.
Considering the fact that this functionality already exists I'm pretty sure this will never be fixed.
So it's just a question of keeping this ticket open (since the ticket has some sense) and never done or just close it.

**Comment by mvysny** [comment]
There are the following arguments for the benefit of having such API:

1. The `Component` itself implements both `AttachNotifier` and `DetachNotifier` but it doesn't offer any means of checking whether the component is attached. The current API seems to be lacking; adding `isAttached()` function would logically complete the API.
2. The `Element.getNode()` API is marked "This method is meant for internal use only." in javadoc, and so I'm not even supposed to call this API.
3. The `StateNode.isAttached()` is very low-level: I, as a Vaadin programmer, have no idea what the state tree is and what it means to be attached to one.

Additional remarks:

- the 'being attached' state is not well defined: e.g. is the component attached when the UI is closed? This is something that would be good to document in `AttachNotifier` and `DetachNotifier`.

I therefore insist on keeping the bug open as a minimum; moreover I think the API would be beneficial and should be implemented.

**Comment by denis-anisimov** [comment]
Alright. The ticket makes sense.
Even though I disagree with it, so let's keep it open.

**Comment by mvysny** [comment]
Thank you.

The `Component::getElement::getNode::isAttached` returns true during the `onDetach()` function / detach listener call. Is there a way to return false in such place please?

**Comment by denis-anisimov** [comment]
I'm not sure.......
This is how the code is written:
```
onDetach();
this.parent = parent;
```

Technically you are right: the parent is assigned only after the detach event is fired but should be assigned before.
On the other hand there can be a reason for that and changing the behavior may cause serious regressions.

Also you don't need to check `isAttached` inside on detach : you know that the semantic of the method is : detached.
But I understand that in practice you have an extracted method which relies normally only on one state based on `isAttached` instead of several info sources so it's inconvenient.
Anyway, this is another issue . Please create another ticket.

**Comment by mvysny** [comment]
> But I understand that in practice you have an extracted method which relies normally only on one state based on isAttached instead of several info sources so it's inconvenient.

I agree, exactly my thoughts.

> Anyway, this is another issue . Please create another ticket.

I agree, I've opened https://github.com/vaadin/flow/issues/8809 and tried to summarize the discussion above there.

**Comment by mvysny** [comment]
Will this be backported to Vaadin 14? (it's marked as released but I can't see the function in Vaadin 14.6.8).

**Comment by pleku** [comment]
Hasn't been backported because nobody has just not done it. It should have been though, so I'm even willing to make another Flow 2.7 beta for it.

**Comment by mvysny** [comment]
Thank you @pleku , really appreciated :1st_place_medal:
